src/app/clusters/operational-credentials-server/operational-credentials-server.cpp

```diff
@@ -417,6 +424,135 @@ bool emberAfOperationalCredentialsClusterUpdateNOCCallback(EndpointId endpoint,
+    }
+    else
+    {
+        SuccessOrExit(status = EMBER_ZCL_STATUS_FAILURE);
+    }
+    SuccessOrExit(err = writer->Put(TLV::ContextTag(0), derBufSpan));
+    SuccessOrExit(err = commandObj->FinishCommand());
+
+exit:
+    if (status == EMBER_ZCL_STATUS_FAILURE)
+    {
+        emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: Failed CertChainRequest.");
+        emberAfSendImmediateDefaultResponse(status);
+    }
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(Zcl, "Failed to encode response command: %s", ErrorStr(err));
+    }
+
+    return true;
+}
+
+bool emberAfOperationalCredentialsClusterAttestationRequestCallback(EndpointId endpoint, app::CommandHandler * commandObj,
+                                                                    ByteSpan attestationNonce)
+{
+    EmberAfStatus status    = EMBER_ZCL_STATUS_SUCCESS;
+    CHIP_ERROR err          = CHIP_NO_ERROR;
+    TLV::TLVWriter * writer = nullptr;
+    Platform::ScopedMemoryBuffer<uint8_t> attestationElements;
+    size_t attestationElementsLen;
+    Crypto::P256ECDSASignature signature;
+    PeerConnectionState * state;
+    uint8_t md[Crypto::kSHA256_Hash_Length];
+    MutableByteSpan messageDigestSpan(md);
+
+    emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: commissioner has requested Attestation");
+
+    app::CommandPathParams cmdParams = { emberAfCurrentEndpoint(), /* group id */ 0, ZCL_OPERATIONAL_CREDENTIALS_CLUSTER_ID,
+                                         ZCL_ATTESTATION_RESPONSE_COMMAND_ID, (app::CommandPathFlags::kEndpointIdValid) };
+
+    // TODO: remove line below
+    Credentials::SetDeviceAttestationCredentialsProvider(Credentials::Examples::GetExampleDACProvider());
+    Credentials::DeviceAttestationCredentialsProvider * dacProvider = Credentials::GetDeviceAttestationCredentialsProvider();
+
+    VerifyOrExit(commandObj != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrExit(attestationNonce.size() == 32, status = EMBER_ZCL_STATUS_FAILURE);
+
+    {
+        uint8_t certDeclBuf[512];
+        MutableByteSpan certDeclSpan(certDeclBuf);
+
+        SuccessOrExit(err = dacProvider->GetCertificationDeclaration(certDeclSpan));
+
+        attestationElementsLen = certDeclSpan.size() + attestationNonce.size() + sizeof(uint64_t) * 8;
+        VerifyOrExit(attestationElements.Alloc(attestationElementsLen), err = CHIP_ERROR_NO_MEMORY);
+
+        MutableByteSpan attestationElementsSpan(attestationElements.Get(), attestationElementsLen);
+        SuccessOrExit(err = Credentials::ConstructAttestationElements(certDeclSpan, attestationNonce, 0, ByteSpan(), nullptr, 0, 0,
+                                                                      0, attestationElementsSpan));
+        attestationElementsLen = attestationElementsSpan.size();
+    }
+
+    // Retrieve attestation challenge
+    state = commandObj->GetExchangeContext()->GetExchangeMgr()->GetSessionMgr()->GetPeerConnectionState(
+        commandObj->GetExchangeContext()->GetSecureSession());
```

Comment:
There should be a utility on `app::CommandHandler` to get the `PeerConnectionState` rather than needing this boilerplate a few places across the codebase.

Reply:
seems to be the only way to do it (as is)

Reply:
Yes, I was saying we need to replace this boilerpltate. If you can add a TODO it would help

Reply:
Later you do:

```
state->GetSecureSession().GetAttestationChallenge()
```

Could you just avoid this altogether and replace:
```
state->GetSecureSession().GetAttestationChallenge()
```

with `commandObj->GetExchangeContext()->GetSecureSession().GetAttestationChallenge()` ?